### PR TITLE
no_timer option can work in non debug mode

### PR DESCRIPTION
**Pull request by wangrzneu**
remove debug_mode limit of no_timer option

### CODE REVIEW

**Comment by wangrzneu** [comment]
PLTA @thibaultcha

**Comment by rainingmaster** [comment]
maybe you can add some test case, so we can know the effect?

File: lib/resty/upstream/healthcheck.lua [hunk 1]
```diff
@@ -54,7 +54,6 @@ local function errlog(...)
 end
 
 local function debug(...)
-    -- print("debug mode: ", debug_mode)
```
Comment (doujiang24) [inline lib/resty/upstream/healthcheck.lua:57]: better not introduce unrelated changes.

**Comment by doujiang24** [comment]
@wangrzneu will you take care of the Travis failure thing?
https://travis-ci.org/github/openresty/lua-resty-upstream-healthcheck/jobs/670113565?utm_medium=notification&utm_source=github_status

**Comment by wangrzneu** [comment]
> @wangrzneu will you take care of the Travis failure thing?
> https://travis-ci.org/github/openresty/lua-resty-upstream-healthcheck/jobs/670113565?utm_medium=notification&utm_source=github_status

Yes, I add a new case and I will try to make it pass.

**Comment by wangrzneu** [comment]
> maybe you can add some test case, so we can know the effect?

Done. I add `TEST 16`.

**Comment by wangrzneu** [comment]
> @wangrzneu will you take care of the Travis failure thing?
> https://travis-ci.org/github/openresty/lua-resty-upstream-healthcheck/jobs/670113565?utm_medium=notification&utm_source=github_status

I have fixed it. But something wrong happens in travis CI.

File: lib/resty/upstream/healthcheck.lua (hunk 1 above)
Comment (wangrzneu) [reply lib/resty/upstream/healthcheck.lua]: I have reverted these unrelated changes.

**Comment by wangrzneu** [comment]
> > @wangrzneu will you take care of the Travis failure thing?
> > https://travis-ci.org/github/openresty/lua-resty-upstream-healthcheck/jobs/670113565?utm_medium=notification&utm_source=github_status
> 
> I have fixed it. But something wrong happens in travis CI.

I fix CI by updating OPENSSL_VER.

**Comment by doujiang24** [comment]
@wangrzneu 
can you explain why you need this change?

**Comment by rainingmaster** [comment]
I think the no_timer should only run in the debug mode,  otherwise the health check will run one time only,  and it can not check the issue of upsteams, right?

**Comment by wangrzneu** [comment]
> @wangrzneu
> can you explain why you need this change?

This change will help me change the health check options without reloading the openresty process.

And the health check can be triggered using a timer outside the openresty process.

**Comment by wangrzneu** [comment]
> I think the no_timer should only run in the debug mode, otherwise the health check will run one time only, and it can not check the issue of upsteams, right?

The health check can be triggered using a timer outside the openresty process.

**Comment by doujiang24** [comment]
> And the health check can be triggered using a timer outside the openresty process.

@wangrzneu can you give a sample that how you use it? I don't think I got you.

**Comment by wangrzneu** [comment]
> > And the health check can be triggered using a timer outside the openresty process.
> 
> @wangrzneu can you give a sample that how you use it? I don't think I got you.


My nginx.conf is like: 

```
worker_processes  1;
error_log logs/error.log warn;

events {
    worker_connections 1024;
}

http {
    upstream foo.com {
        server 127.0.0.1:12345;
    }

    server {
        listen 12345;

        location = /status {
            content_by_lua_block {
                ngx.say("ok")
            }
        }
        location / {
            content_by_lua_block {
                ngx.say("foo.com")
            }
        }
    }
    lua_shared_dict healthcheck 1m;
    server {
        listen 8080;
        location / {
            proxy_pass http://foo.com;
        }
    }

    server {
        listen 9090;
        location /healthcheck {
            content_by_lua_block {
                local hc = require "resty.upstream.healthcheck"
                local args = ngx.req.get_uri_args()
                local ok, err = hc.spawn_checker{
                    shm = "healthcheck",
                    upstream = args["upstream"],
                    type = args["type"],
                    http_req = args["probe"] .. " HTTP/1.0\r\nHost: localhost\r\n\r\n",
                    timeout = tonumber(args["timeout"]),
                    fall = tonumber(args["fail"]),
                    no_timer = true,
                }
                if not ok then
                    ngx.log(ngx.ERR, "failed to spawn health checker: ", err)
                    return
                end
                ngx.say(hc.status_page()) 
            }
        }
    }
}
```

I use another program (implemented by golang) to call the /healthcheck periodically, and the arguments may be changed.

File: t/sanity.t [hunk 2]
```diff
@@ -1451,3 +1451,82 @@ healthcheck: peer 127\.0\.0\.1:12355 was checked to be not ok
 healthcheck: peer 127\.0\.0\.1:12356 was checked to be not ok
 healthcheck: peer 127\.0\.0\.1:12359 was checked to be not ok
 $/
+
+=== TEST 16: effect of no_timer (healthcheck options can be changed)
```
Comment (doujiang24) [inline t/sanity.t:1455]: style: we need three blank lines before this line. you are recommended to use the `reindex` tool to do it automatically:
https://github.com/openresty/openresty-devel-utils/blob/master/reindex

doc is still in the PR: https://github.com/openresty/openresty-devel-utils/pull/29/files#diff-04c6e90faac2675aa89e2176d2eec7d8R237

**Comment by doujiang24** [comment]
@wangrzneu I think I got you now.

my opinion, it should be ok to merge it into master though it's not the recommended way to use this library.
this PR is a small change and it really helps @wangrzneu, so I vote for it.
@thibaultcha @spacewander @rainingmaster thoughts?

File: t/sanity.t (hunk 2 above)
Comment (wangrzneu) [reply t/sanity.t:1457]: done

**Comment by spacewander** [comment]
`no_timer` is confusing for produce, maybe we can rename it as `once`?

**Comment by rainingmaster** [comment]
@doujiang24 I agree with you, the change is small and no impact on existing functions and parameters. But maybe the [test cases](https://github.com/openresty/lua-resty-upstream-healthcheck/pull/68/files#diff-900f41bd5ef6ff9448f79dfff3418b2eR1457) should be more simple and  check `no_timer`(or `once` is better) only? Because [test case like here](https://github.com/openresty/lua-resty-upstream-healthcheck/pull/68/files#diff-900f41bd5ef6ff9448f79dfff3418b2eR1481) maybe make other confuse.

**Comment by rainingmaster** [comment]
@wangrzneu I think maybe you should test the  case when has [multiple upstream](https://github.com/openresty/lua-resty-upstream-healthcheck#multiple-upstreams), because [upstream_checker_statuses](https://github.com/openresty/lua-resty-upstream-healthcheck/blob/master/lib/resty/upstream/healthcheck.lua#L46) is init in content_by_lua_block, so maybe some worker do not have the specific key in `upstream_checker_statuses`.

In first request, check with foo.com, hit 0 worker:
```
Upstream foo.com
    Primary Peers
        127.0.0.1:12354 up
    Backup Peers

Upstream boo.com (NO checkers)
    Primary Peers
        127.0.0.1:12354 up
    Backup Peers
```

In second request, check with boo.com, hit 1 worker:
```
Upstream foo.com (NO checkers)
    Primary Peers
        127.0.0.1:12354 up
    Backup Peers

Upstream boo.com
    Primary Peers
        127.0.0.1:12354 up
    Backup Peers
```

In addition, welcome to the official form https://forum.openresty.us/ to share and discuss
